{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=6> arraylize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "def convert_sequences_to_array(sequences):\n",
    "    '''\n",
    "    inputs: sequence of nucleotides represented as a string composed of A, C, G, T\n",
    "    outputs: a list of numpy array representations of a sequence with:\n",
    "             A = [1, 0, 0, 0]\n",
    "             C = [0, 1, 0, 0]\n",
    "             G = [0, 0, 1, 0]\n",
    "             T = [0, 0, 0, 1]\n",
    "             \n",
    "    '''\n",
    "\n",
    "    nucleotide_array_dict = {'A': [1, 0, 0, 0],\n",
    "                             'C': [0, 1, 0, 0],\n",
    "                             'G': [0, 0, 1, 0],\n",
    "                             'T': [0, 0, 0, 1],\n",
    "                             'N': [0.25,0.25,0.25,0.25],\n",
    "                             'R':[0.5,0,0.5,0],\n",
    "                             'Y':[0,0.5,0,0.5],\n",
    "                             'M':[0.5,0.5,0,0],\n",
    "                             'K':[0,0,0.5,0.5],\n",
    "                             'S':[0,0.5,0.5,0],\n",
    "                             'W':[0.5,0,0,0.5],\n",
    "                             'B':[0,0.33,0.33,0.33],\n",
    "                             'D':[0.33,0,0.33,0.33],\n",
    "                             'H':[0.33,0.33,0,0.33],\n",
    "                             'V':[0.33,0.33,0.33,0]}\n",
    "    \n",
    "    sequence_array_list = []\n",
    "    for seq in sequences:\n",
    "        seq_array = []\n",
    "        for nuc in seq:\n",
    "            seq_array.append(nucleotide_array_dict[nuc])\n",
    "        seq_array = np.array(seq_array) \n",
    "        sequence_array_list.append(seq_array)\n",
    "    return sequence_array_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=6> Get TATA_TSS,nonTATA_TSS,TATA_nonTSS,nonTATA_nonTSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "def get_TSS_nonTSS(gff_name,fasta_name):   \n",
    "    rdata=[]\n",
    "    comment=[]\n",
    "    delnum=0\n",
    "    with open(gff_name,\"r\") as f:\n",
    "        rdata=f.readlines()\n",
    "    for i in range(9):\n",
    "        del(rdata[0])\n",
    "    for i in range(len(rdata)):\n",
    "        if rdata[i][0]==\"#\":\n",
    "            comment.append(i)\n",
    "        rdata[i]=rdata[i].split()\n",
    "    for i in range(len(comment)):\n",
    "        del(rdata[comment[i]-delnum])\n",
    "        delnum+=1\n",
    "    isgene=[]\n",
    "    for i in range(len(rdata)):\n",
    "        if rdata[i][2]==\"gene\":\n",
    "            isgene.append(rdata[i])\n",
    "    typelist=[]\n",
    "    for i in range(len(rdata)):\n",
    "        if rdata[i][2] not in typelist:\n",
    "            typelist.append(rdata[i][2])\n",
    "    genetrandp=[]\n",
    "    genetrandn=[]\n",
    "    genetrandun=[]\n",
    "    for i in range (len(isgene)):\n",
    "        if isgene[i][6]==\"+\":\n",
    "            genetrandp.append([isgene[i][0],isgene[i][3],isgene[i][4],isgene[i][6]])\n",
    "        elif isgene[i][6]==\"-\":\n",
    "            genetrandn.append([isgene[i][0],isgene[i][3],isgene[i][4],isgene[i][6]])\n",
    "        else:\n",
    "            genetrandun.append([isgene[i][0],isgene[i][3],isgene[i][4],isgene[i][6]])\n",
    "\n",
    "    TSSlocation=[]\n",
    "\n",
    "    for i in range(len(genetrandp)):\n",
    "        TSSlocation.append([genetrandp[i][0],int(genetrandp[i][1])-50,int(genetrandp[i][1])+200])\n",
    "    rfasta=[]\n",
    "    fasta2=[]\n",
    "    genename=[]\n",
    "    string=\"\"\n",
    "    with open(fasta_name,\"r\") as f:\n",
    "        rfasta=f.readlines()\n",
    "    for i in range(len(rfasta)):\n",
    "        if rfasta[i][0]==\">\":\n",
    "            fasta2.append(string)\n",
    "            string=\"\"\n",
    "            genename.append(rfasta[i])\n",
    "        else:\n",
    "            string+=rfasta[i].rstrip()\n",
    "    del(fasta2[0])\n",
    "    del(genename[-1]) \n",
    "    for i in range(len(genename)):\n",
    "        genename[i]=genename[i].split()[0].lstrip(\">\")\n",
    "    genefasta={}\n",
    "    for i in range(len(genename)):\n",
    "        genefasta.update({genename[i]:fasta2[i]})\n",
    "    TSSfasta=[]\n",
    "    for i in range(len(TSSlocation)):\n",
    "        if (TSSlocation[i][0] in genefasta.keys()):\n",
    "            TSSfasta.append(genefasta[TSSlocation[i][0]][TSSlocation[i][1]:TSSlocation[i][2]].upper())\n",
    "    n=0        \n",
    "    while(n<len(TSSfasta)):\n",
    "        while(len(TSSfasta[n])!=250):\n",
    "            del(TSSfasta[n])\n",
    "        n+=1\n",
    "    nonTSSstart=[]\n",
    "    nonTSSname=[]\n",
    "    for i in range(50000):\n",
    "        temp=random.randint(0,len(genetrandp)-1)\n",
    "        if int(genetrandp[temp][2])-250>int(genetrandp[temp][1]):\n",
    "            nonTSSname.append(temp)\n",
    "            nonTSSstart.append(random.randint(int(genetrandp[temp][1]),int(genetrandp[temp][2])-250))\n",
    "    nonTSSfasta=[]           \n",
    "    for i in range(len(nonTSSname)):\n",
    "        if  genetrandp[nonTSSname[i]][0] in genefasta.keys():\n",
    "            name=genetrandp[nonTSSname[i]][0]\n",
    "            nonTSSfasta.append(genefasta[name][nonTSSstart[i]:nonTSSstart[i]+250].upper())\n",
    "    TATA_TSSfasta=[]\n",
    "    TATA_nonTSSfasta=[]\n",
    "    for i in range(len(TSSfasta)):\n",
    "        for j in range(45):\n",
    "            if TSSfasta[i][j:j+4]==\"TATA\":\n",
    "                TATA_TSSfasta.append(TSSfasta[i])\n",
    "    for i in range(len(nonTSSfasta)):\n",
    "        for j in range(45):\n",
    "            if nonTSSfasta[i][j:j+4]==\"TATA\":\n",
    "                TATA_nonTSSfasta.append(nonTSSfasta[i])\n",
    "    nonTATA_TSSfasta=[]\n",
    "    nonTATA_nonTSSfasta=[]\n",
    "    for i in range(len(TSSfasta)):\n",
    "        if TSSfasta[i] not in TATA_TSSfasta:\n",
    "            nonTATA_TSSfasta.append(TSSfasta[i])\n",
    "    for i in range(len(nonTSSfasta)):\n",
    "        if nonTSSfasta[i] not in TATA_nonTSSfasta:\n",
    "            nonTATA_nonTSSfasta.append(nonTSSfasta[i])\n",
    "    print(len(nonTATA_nonTSSfasta))\n",
    "    TATA_TSSinput=np.array(convert_sequences_to_array(TATA_TSSfasta))\n",
    "    nonTATA_TSSinput=np.array(convert_sequences_to_array(nonTATA_TSSfasta))\n",
    "    TATA_nonTSSinput=np.array(convert_sequences_to_array(TATA_nonTSSfasta))\n",
    "    nonTATA_nonTSSinput=np.array(convert_sequences_to_array(nonTATA_nonTSSfasta))\n",
    "    return TATA_TSSinput,nonTATA_TSSinput,TATA_nonTSSinput,nonTATA_nonTSSinput"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TATA_TSS,nonTATA_TSS,TATA_nonTSS,nonTATA_nonTSS=\\\n",
    "# get_TSS_nonTSS(gff_name=\"GCF_000001405.39_GRCh38.p13_genomic.gff\",fasta_name=\"GCF_000001405.39_GRCh38.p13_genomic.fna\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "39813"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=6> Combine and labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def TATA_nonTATA_data_labels(TATA_TSS,nonTATA_TSS,TATA_nonTSS,nonTATA_nonTSS):\n",
    "    TATAinput=list(TATA_TSS)+list(TATA_nonTSS)\n",
    "    nonTATAinput=list(nonTATA_TSS)+list(nonTATA_nonTSS)\n",
    "    TATAlabels=[0]*len(TATA_TSS)+[1]*len(TATA_nonTSS)\n",
    "    TATAlabels=np.array(TATAlabels)\n",
    "    nonTATAlabels=[0]*len(nonTATA_TSS)+[1]*len(nonTATA_nonTSS)\n",
    "    nonTATAlabels=np.array(nonTATAlabels)\n",
    "    TATAinput=np.array(TATAinput)\n",
    "    nonTATAinput=np.array(nonTATAinput)\n",
    "    return TATAinput,nonTATAinput,TATAlabels,nonTATAlabels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TATAinput,nonTATAinput,TATAlabels,nonTATAlabels=\\\n",
    "# TATA_nonTATA_data_labels(TATA_TSS=TATA_TSS,nonTATA_TSS=nonTATA_TSS,TATA_nonTSS=TATA_nonTSS,nonTATA_nonTSS=nonTATA_nonTSS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=6> Padding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_len=21\n",
    "motif_num=200\n",
    "pad_array = [0.25,0.25,0.25,0.25]\n",
    "def pad_sequence_arrays(seq_arrays,pad_array, num, direction):\n",
    "    pad_array = [0.25,0.25,0.25,0.25]\n",
    "    padded_seq = []\n",
    "    for i in seq_arrays:\n",
    "        if direction == 'both':\n",
    "            padded_seq.append([pad_array]*num + i.tolist() + [pad_array]*num)\n",
    "        elif direction == 'pre':\n",
    "            padded_seq.append([pad_array]*num + i.tolist())\n",
    "        elif direction == 'post':\n",
    "            padded_seq.append(i.tolist() + [pad_array]*num)\n",
    "        else:\n",
    "            print('Wrong Parameter!!')\n",
    "    return np.array(padded_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# padded_TATA= pad_sequence_arrays(TATAinput, pad_array, motif_len-1, 'both')\n",
    "# padded_nonTATA= pad_sequence_arrays(nonTATAinput, pad_array, motif_len-1, 'both')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=6> Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(gff_name,fasta_name):\n",
    "    TATA_TSS,nonTATA_TSS,TATA_nonTSS,nonTATA_nonTSS=get_TSS_nonTSS(gff_name,fasta_name)\n",
    "    TATAinput,nonTATAinput,TATAlabels,nonTATAlabels=\\\n",
    "    TATA_nonTATA_data_labels(TATA_TSS=TATA_TSS,nonTATA_TSS=nonTATA_TSS,TATA_nonTSS=TATA_nonTSS,nonTATA_nonTSS=nonTATA_nonTSS)\n",
    "    padded_TATA= pad_sequence_arrays(TATAinput, pad_array, motif_len-1, 'both')\n",
    "    padded_nonTATA= pad_sequence_arrays(nonTATAinput, pad_array, motif_len-1, 'both')\n",
    "    return padded_TATA,padded_nonTATA,TATAlabels,nonTATAlabels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39775\n"
     ]
    }
   ],
   "source": [
    "padded_TATA,padded_nonTATA,TATAlabels,nonTATAlabels=\\\n",
    "main(gff_name=\"GCF_000001405.39_GRCh38.p13_genomic.gff\",fasta_name=\"GCF_000001405.39_GRCh38.p13_genomic.fna\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(\"T5\",padded_TATA)\n",
    "np.save(\"nT5\",padded_nonTATA)\n",
    "np.save(\"T5labels\",TATAlabels)\n",
    "np.save(\"nT5labels\",nonTATAlabels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "59247"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(padded_nonTATA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
